File: Formula1API/ServicesImpl/TeamService.cs
using AutoMapper;
using Formula1API.Data;
using Formula1API.Dtos;
using Formula1API.Models;
using Formula1API.Services;
using Microsoft.EntityFrameworkCore;
using NuGet.Packaging;

namespace Formula1API.ServicesImpl
{
    public class TeamService : Exception, ITeamService
    {
        private readonly Formula1DbContext _dbContext;
        private readonly IMapper _mapper;

        public TeamService(Formula1DbContext dbContext, IMapper mapper)
        {
            _dbContext = dbContext;
            _mapper = mapper;
        }

        public async Task<ICollection<Team>> GetAllTeams()
        {
            var team = (await _dbContext.Teams
                .Include(t => t.Drivers)
                .ToListAsync());

            return team.Select(t => _mapper.Map<Team>(t)).ToList();
        }

        public async Task<Team> GetTeam(int id)
        {
            var team = await _dbContext.Teams
                .Include(t => t.Drivers)
                .FirstOrDefaultAsync(t => t.Id == id);

            if (team == null)
            {
                throw new Exception("Team not found");
            }

            return _mapper.Map<Team>(team);
        }

        public async Task<Team> GetTeamOfDriver(int id)
        {
            var driver = await _dbContext.Drivers.FirstOrDefaultAsync(d => d.Id == id);

            if (driver == null)
            {
                throw new Exception("Driver not found");
            }

            var team = await _dbContext.Teams
                .FirstOrDefaultAsync(t => t.Drivers.Any(d => d.Id == driver.Id));

            if (team == null)
            {
                throw new Exception("Team not found");
            }

            return _mapper.Map<Team>(team);
        }

        public async Task<ICollection<Driver>> GetDriversOfTeam(int id)
        {
            var team = await _dbContext.Teams.Include(t => t.Drivers).FirstOrDefaultAsync(t => t.Id == id);

            if (team == null)
            {
                throw new Exception("Team not found");
            }

            return _mapper.Map<ICollection<Driver>>(team.Drivers);
        }

        public async Task<ICollection<Race>> GetRacesOfTeam(int id)
        {
            var team = await _dbContext.Teams.FirstOrDefaultAsync(t => t.Id == id);

            if (team == null)
            {
                throw new Exception("Team not found");
            }

            var races = await _dbContext.Races
                .Include(r => r.WinnerDriver)
                .ThenInclude(d => d.Team)
                .Where(r => r.WinnerDriver.Team.Id == id)
                .ToListAsync();

            return _mapper.Map<ICollection<Race>>(races);
        }

        public async Task<Team> AddTeam(TeamDto teamDto)
        {
            var team = new Team
            {
                Name = teamDto.Name,
                CarName = teamDto.CarName,
                Engine = teamDto.Engine,
                Director = teamDto.Director,
            };

            var drivers = await _dbContext.Drivers.Where(d => teamDto.DriversIds.Contains(d.Id)).ToListAsync();

            team.Drivers.AddRange(drivers);

            _dbContext.Teams.Add(team);
            await _dbContext.SaveChangesAsync();

            return _mapper.Map<Team>(team);
        }

        public async Task<Team> EditTeam(int id, TeamDto teamDto)
        {
            var team = await _dbContext.Teams.FindAsync(id);

            if (team == null)
            {
                throw new Exception("Team not found");
            }

            team.Name = teamDto.Name;
            team.CarName = teamDto.CarName;
            team.Engine = teamDto.Engine;
            team.Director = teamDto.Director;

            var drivers = await _dbContext.Drivers.Where(d => teamDto.DriversIds.Contains(d.Id)).ToListAsync();

            team.Drivers.AddRange(drivers);

            await _dbContext.SaveChangesAsync();

            return _mapper.Map<Team>(team);
        }

        public async Task DeleteTeam(int id)
        {
            var team = await _dbContext.Teams.FindAsync(id);

            if (team == null)
            {
                throw new Exception("Team not found");
            }

            _dbContext.Teams.Remove(team);
            await _dbContext.SaveChangesAsync();
        }
    }
}

File: Formula1Client/src/app/team/team.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from "@angular/common/http";
import { API_URL, API_REQ_HEADER } from '../comunicationConstants';
import { Team } from './team.model';
import { Observable } from 'rxjs';
import { Driver } from '../driver/driver.model';
import { Race } from '../race/race.model';

@Injectable({
  providedIn: 'root'
})
export class TeamService {

  private apiUrl = API_URL + "Team/";

  constructor(private http: HttpClient) { }

  getAllTeams(): Observable<Team[]> {
    return this.http.get<Team[]>(this.apiUrl, {headers: API_REQ_HEADER});
  }

  getTeam(id: number): Observable<Team> {
    return this.http.get<Team>(this.apiUrl + id, {headers: API_REQ_HEADER});
  }

  getTeamOfDriver(id: number): Observable<Team> {
    return this.http.get<Team>(this.apiUrl + "driver/" + id, {headers: API_REQ_HEADER});
  }

  getDriversOfTeam(id: number): Observable<Driver[]> {
    return this.http.get<Driver[]>(this.apiUrl + id + "/drivers", {headers: API_REQ_HEADER});
  }

  getRacesOfTeam(id: number): Observable<Race[]> {
    return this.http.get<Race[]>(this.apiUrl + id + "/races", {headers: API_REQ_HEADER});
  }

  addTeam(team: Team): Observable<Team> {
    return this.http.post<Team>(this.apiUrl, team, {headers: API_REQ_HEADER});
  }

  editTeam(id: number, team: Team): Observable<Team> {
    return this.http.put<Team>(this.apiUrl + id, team, {headers: API_REQ_HEADER});
  }

  deleteTeam(id: number): Observable<Boolean> {
    return this.http.delete<Boolean>(this.apiUrl + id, {headers: API_REQ_HEADER});
  }
}
